### ISSUE TITLE
Add webp support for use with SDL2_image

### ISSUE THREAD

**Issue by Pentarctagon** [ValveSoftware/steam-runtime#475]
In the game Battle for Wesnoth we're using the Steam runtime for our Linux builds, and we'd like to start using the webp format as a replacement for large jpeg and png images we currently have due to improvements such as having smaller file sizes and transparency support, however looking at the latest snapshot at https://repo.steampowered.com/steamrt-images-scout/snapshots/ there is no libwebp listed.

Would it be possible to add libwebp to the runtime and build SDL2 with webp support enabled?

**Comment by smcv** [comment]
Am I correct to think that this is entirely a feature request for how we compile [SDL_image](https://github.com/libsdl-org/SDL_image/) (referred to as libsdl2-image in Debian/SteamRT), and does not affect the [core SDL library](https://github.com/libsdl-org/SDL/) (libsdl2 in Debian/SteamRT)?

**Comment by Pentarctagon** [comment]
Correct.

**Comment by smcv** [comment]
This change is making its way through the system. Today's beta of `SteamLinuxRuntime_soldier` has libwebp enabled in its `SDL_image`, so if you have unsupported development builds of your game, you can try this out already:

1. enable the `Steam Linux Runtime` steam compatibility tool for your game
2. the `Steam Linux Runtime` and `Steam Linux Runtime - soldier` compatibility tools should both auto-download
3. select the `client_beta` branch of the `Steam Linux Runtime - soldier` compat tool (officially this is only meant to be used with the Steam Client beta, but right now it should be OK with either the general-availability or beta client)

However, this is not safe to rely on in a published game until the general-availability Steam client is shipping it in `~/.steam/root/ubuntu12_32/steam-runtime`, and I cannot predict how long that will take. I'll let you know when there's more information available.

**Comment by Pentarctagon** [comment]
Sounds great, thanks!

**Comment by smcv** [comment]
This is in Steam beta clients since <https://steamcommunity.com/groups/SteamClientBeta/announcements/detail/3114799541524949802>. It is still not safe to rely on in a published game (I would guess that won't happen until mid January at the earliest).

**Comment by smcv** [comment]
This has been in the General Availability branch of Steam since https://store.steampowered.com/oldnews/102669.

The `Steam Linux Runtime` compatibility tool is currently lagging behind that, and does not have libwebp enabled unless you are opted-in to the `client_beta` branch of `Steam Linux Runtime - soldier`. It should go to the default branch in the next update.

**Comment by smcv** [comment]
> The Steam Linux Runtime compatibility tool is currently lagging behind that, and does not have libwebp enabled unless you are opted-in to the client_beta branch of Steam Linux Runtime - soldier. It should go to the default branch in the next update.

This has now caught up, so all supported environments should now have libwebp and a SDL2_image that uses it. @Pentarctagon, please could you try it out?

**Comment by Pentarctagon** [comment]
It looks like it's working in the actual Steam client - I replaced a .png image with a .webp image and Wesnoth displayed it - however a simple test program to check for webp support when building isn't working when run in a docker image based on the latest Scout snapshot, so currently trying to figure out what's going on there.

**Comment by smcv** [comment]
For a Docker environment, you want `BUILD_ID="0.20211207.0"` or later in `/etc/os-release`. `podman pull registry.gitlab.steamos.cloud/steamrt/scout/sdk:latest` currently gives me 0.20211207.0.

However, I think something might have gone wrong with the lazy-loading mechanism in SDL_image: our container has `/usr/lib/x86_64-linux-gnu/libwebp.so.6`, but `strings /usr/lib/x86_64-linux-gnu/libSDL2_image-2.0.so.0` only mentions `libwebp.so.2` (and there is no hard dependency via `DT_NEEDED` because SDL_image uses `dlopen()`). I'll look into it.

**Comment by smcv** [comment]
Sorry, looks like a build-time dependency wasn't tight enough: it worked in my test environment, but not in final builds. I'll have another try at this in the next Steam beta cycle.

**Comment by smcv** [comment]
For now, you can use `SteamLinuxRuntime_soldier` (as described above) if you want to try this out for your own builds.

**Comment by smcv** [comment]
[Steam beta update 2022-02-01](https://steamcommunity.com/groups/SteamClientBeta/announcements/detail/4515423468952376216) includes a new runtime, scout 0.20220119.0, with a new build of `libsdl2-image` that supports webp using `libwebp.so.6` (really this time). The matching SDK is `registry.gitlab.steamos.cloud/steamrt/scout/sdk:beta`.

As before, this is not yet in a general-availability release.

```
user@host:~$ file ~/tmp/demo.webp
/home/user/tmp/demo.webp: RIFF (little-endian) data, Web/P image, VP8 encoding, 550x368, Scaling: [none]x[none], YUV color, decoders should clamp
user@host:~$ ldd ~/tmp/showimage-scout | grep SDL
	libSDL2_image-2.0.so.0 => not found
	libSDL2-2.0.so.0 => not found
user@host:~$ ~/.steam/root/ubuntu12_32/steam-runtime/run.sh -- env debian_chroot=steam-runtime bash
(steam-runtime)user@host:~$ ldd ~/tmp/showimage-scout |grep SDL
	libSDL2_image-2.0.so.0 => /home/user/.steam/root/ubuntu12_32/steam-runtime/usr/lib/x86_64-linux-gnu/libSDL2_image-2.0.so.0 (0x00007f4e704fe000)
	libSDL2-2.0.so.0 => /home/user/.steam/root/ubuntu12_32/steam-runtime/usr/lib/x86_64-linux-gnu/libSDL2-2.0.so.0 (0x00007f4e70173000)
(steam-runtime)user@host:~$ ~/tmp/showimage-scout ~/tmp/demo.webp
(image is displayed)
```

My test image was https://www.gstatic.com/webp/gallery/1.webp, and the demo program is https://github.com/libsdl-org/SDL_image/blob/main/showimage.c compiled on scout.

**Comment by Pentarctagon** [comment]
It looks like it's working as expected now, thanks!

For the demo program you linked though, it doesn't look like it mentions webp anywhere in it?

**Comment by smcv** [comment]
> For the demo program you linked though, it doesn't look like it mentions webp anywhere in it?

Not explicitly, but it uses `IMG_Load()`, which will detect and load any format supported by SDL_image (or, conversely, won't detect and load images that it doesn't have support for). I passed a webp image as a command-line parameter, so the only way it can succeed in displaying that is by supporting webp :-)

**Comment by kisak-valve** [comment]
Closing as implemented.

**Comment by Pentarctagon** [comment]
Just to confirm - this is now generally available, correct?

**Comment by kisak-valve** [comment]
Rechecking the mainline release channel, which is dated 2022-01-20, I may have jumped the gun here. It looks like it's before the most recent adjustment so I'll re-open this for now, even though it looks like there's nothing left to be done here except wait for the next time Steam's beta gets promoted to the mainline release. Thanks for the nudge.

**Comment by smcv** [comment]
How to tell whether this has arrived: `~/.steam/root/ubuntu12_32/steam-runtime/version.txt` will say 0.20220119.0 or later, and https://repo.steampowered.com/steamrt-images-scout/snapshots/latest-steam-client-general-availability/VERSION.txt will catch up with that version within about a day (there's a periodic check, but it lags behind a bit, because it's safer for the SDK to be too old than for it to be too new).

This is not *necessarily* connected to the release date of the Steam client, because there are sometimes new builds of the GA Steam client without copying the new runtime from the beta branch, for example if there's a hotfix update to fix an important bug.

**Comment by Pentarctagon** [comment]
Nice, it looks like it's there now. Thanks!

**Comment by smcv** [comment]
> Nice, it looks like it's there now. Thanks!

Fixed in Steam client 2022-02-05 (1646446125).
